modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Node.scala

```diff
@@ -0,0 +1,542 @@
+      dataType: AbsoluteIri,
+      languageTag: Option[LanguageTag] = None
+  ) extends Node {
+    override def isBlank: Boolean           = false
+    override def isIri: Boolean             = false
+    override def isLiteral: Boolean         = true
+    override def asBlank: Option[BNode]     = None
+    override def asIri: Option[IriNode]     = None
+    override def asLiteral: Option[Literal] = Some(this)
+
+    /**
+      * @return true if the value is numeric, false otherwise
+      */
+    def isNumeric: Boolean =
+      xsd.numericDataTypes.contains(dataType)
+
+    /**
+      * @return true if the value is a string or a string tagged with a language, false otherwise
+      */
+    def isString: Boolean =
+      dataType == xsd.string || dataType == rdfsyntax.langString
+  }
+
+  object Literal {
+    object xsd {
+      private final val schema = Url.unsafe("http://www.w3.org/2001/XMLSchema")
+
+      private final implicit def fragment(string: String): Fragment = new Fragment(string)
+
+      final val string = schema.withFragment("string")
+
+      final val boolean = schema.withFragment("boolean")
+
+      final val byte    = schema.withFragment("byte")
+      final val short   = schema.withFragment("short")
+      final val int     = schema.withFragment("int")
+      final val integer = schema.withFragment("integer")
+      final val long    = schema.withFragment("long")
+      final val decimal = schema.withFragment("decimal")
+
+      final val negativeInteger    = schema.withFragment("negativeInteger")
+      final val nonNegativeInteger = schema.withFragment("nonNegativeInteger")
+      final val nonPositiveInteger = schema.withFragment("nonPositiveInteger")
+      final val positiveInteger    = schema.withFragment("positiveInteger")
+      final val unsignedByte       = schema.withFragment("unsignedByte")
+      final val unsignedShort      = schema.withFragment("unsignedShort")
+      final val unsignedInt        = schema.withFragment("unsignedInt")
+      final val unsignedLong       = schema.withFragment("unsignedLong")
+
+      // format: off
+      final val numericDataTypes: Set[AbsoluteIri] = Set(
+        byte, short, int, integer, long, decimal,
+        negativeInteger, nonNegativeInteger, nonPositiveInteger, positiveInteger,
+        unsignedByte, unsignedShort, unsignedInt, unsignedLong)
+      // format: on
+    }
+    object rdfsyntax {
+      final val langString = Url.unsafe("http://www.w3.org/1999/02/22-rdf-syntax-ns#langString")
+    }
+
+    /**
+      * Creates a new Literal node from the arguments.
+      *
+      * @param lexicalForm the literal lexical form
+      * @param dataType    the data type of the literal
+      */
+    final def apply(lexicalForm: String, dataType: AbsoluteIri): Literal =
+      new Literal(lexicalForm, dataType)
+
+    /**
+      * Creates a new string literal.  String literals have the ''http://www.w3.org/2001/XMLSchema#string'' data type.
+      *
```

Comment:
what if instead of all those constructors you would use shapeless:
```scala
type SchemaTypes = String :+: Long :+: Boolean :+: Byte :+: Short :+: ... :+: CNil
  private object toSchema extends Poly1 {
    implicit def caseString      = at[String](_ => schema.withFragment("string"))
    implicit def caseBoolean      = at[String](_ => schema.withFragment("boolean"))
   ...
  }
final def apply[A](value: A)(implicit I: Inject[SchemaTypes, A]) {
new Literal(value.toString, value.map(toSchema))
}
```

Reply:
I don't think it will reduce the boilerplate because you still need to case for each type... In fact it may be even longer :)

For this specific case I don't think it's worth adding shapeless types.

Reply:
well, it requires just one constructor instead of one per each type defined in `xsd`. And the `Poly1` block is the same which you already have where you create all your `schema.withFragment("boolean")` `schema.withFragment("string")`

But yes, it isn't required :)

Reply:
It may be interesting to still consider it for other well known literal types, i.e. date, time, datetime etc.

Reply:
thinking about it some more, maybe a typeclass would be even better...

```
trait LiteralValue[A] {
  def literal(a: A): Literal
  def lexicalForm(a: A): String
  def dataType(a: A): AbsoluteIri
}

implicit val intValue: LiteralValue[Int] = new LiteralValue[Int] {
   //...
}
```